llvm/lib/Transforms/Scalar/LoopLoadElimination.cpp

```diff
@@ -598,10 +598,11 @@ class LoadEliminationForLoop {
       }
 
       // Point of no-return, start the transformation.  First, version the loop
-      // if necessary.
-
-      LoopVersioning LV(LAI, Checks, L, LI, DT, PSE.getSE());
-      LV.versionLoop();
+      // if it's not a single-iteration loop.
+      if (!PSE.getBackedgeTakenCount()->isOne()) {
```

Comment:
Shouldn't this bail out instead of just skipping versioning?

I think the code below assumes versioning happened and it may be required for correctness. If we don't version, we would perform LLE on a single iteration loop without performing the required versioning?

Reply:
Actually, that's a good point: while I agree that versioning is required in the general case for correctness, I'm not sure if versioning a single-iteration loop is required for correctness. Doesn't versioning, after all, create a version of the loop that is guaranteed to execute once? If the loop executes just once anyway, will the transformed loop possibly not execute after LLE? I'm confused because the test I included seems to be a correct transformation: is this just incidental?

Reply:
Here's the Alive2 proof for the transformation in the test: https://alive2.llvm.org/ce/z/fKWdPi.

Reply:
That's an interesting point, thanks! One thing to note is that BTC == 1 means that the backed I taken once, i.e. the loop body executes twice IIUC (e.g. see https://llvm.godbolt.org/z/E385vqb4M for `@single_iteration_unknown_stride`, which has BTC = 0)

Taking a step back, `@lver.check.unnecessary` from https://github.com/llvm/llvm-project/issues/96656 should be handled by checking for BTC == 0?

For BTC == 0, it should indeed not be a correctness issue, as we only execute a single iteration. But then the transform isn't needed, so it would still be better to bail out.

> Here's the Alive2 proof for the transformation in the test: https://alive2.llvm.org/ce/z/fKWdPi.

Thanks for sharing. Unfortunately Alive2's reasoning about loops can be a bit surprising. I am not able to summarize the way the reasoning works at the moment, but IIUC it only reasons about the straight-line paths in the CFG, so for the link you shared only about the first loop iteration.

To work around the limitation, there are the `-src-unroll` and `-tgt-unroll` options. E.g. when unrolling by 2, verification fails because now we verify both the first and second iteration. (Apologies, my explanation here may not be completely precise)

https://alive2.llvm.org/ce/z/7g7sg-

Reply:
Sorry about this huge confusion and mess. It looks like, with a single iteration, LLE doesn't version the loop. I think the regression is observed when BTC = 1. Investigating further.

Reply:
Okay, so the root cause isn't BTC = 1, since LLE does operate on loops with BTC = 1, after versioning them. The root cause is that certain dependencies that were previously unknown (in this case `findStoreToLoadDependences` returns no candidates, and LLE bails out before versioning) are now correctly analyzed.  LLE is actually more powerful in this case, and I'm tempted to classify the regression in #96656 as invalid. Will update that bug with all the information and close this PR.

Again, sorry for the confusion. I got confused from the different outputs of llvm-reduce with loop-versioning and loop-load-elim in the pipeline.